### PR TITLE
prov/sockets: Fix #3217, bug in buffered receive path

### LINKED ISSUE #3217: Sockets: The provider is losing messages when FI_MULTI_RECV space runs low.
First off, this issue was reported by LBNL when testing GASNet and is described in detail on their Bugzilla here: https://upc-bugs.lbl.gov/bugzilla/show_bug.cgi?id=3597

Second, here are the hints being used in GASNet:
```
hints->caps                   = FI_RMA | FI_MSG | FI_MULTI_RECV;
hints->mode                   = FI_CONTEXT; 
hints->addr_format  = FI_FORMAT_UNSPEC;
hints->tx_attr->op_flags = FI_DELIVERY_COMPLETE;
hints->ep_attr->type = FI_EP_RDM;
hints->domain_attr->threading = FI_THREAD_DOMAIN;
hints->domain_attr->control_progress  = FI_PROGRESS_MANUAL;
hints->domain_attr->resource_mgmt  = FI_RM_ENABLED;
hints->domain_attr->av_type = FI_AV_TABLE;
```
Also note that this test is running in a single thread.

The test that is hanging due to lost messages is testam, which fires a large number of messages at once. This failure is sporadic and occurs at different times. Sometimes it takes place in the test warmup, sometimes it takes place in the test body, sometimes it happens after the test is run. To run the test and reproduce the failure, I loop the test as so: 
`$ while FI_PROVIDER=sockets ./contrib/gasnetrun_ofi -np 2 ./testam 10000 8192 F; do :; done `

The specific test above (designated by specifying 'F') is a two-way flood test, meaning both nodes are sending messages to each other. On examining the processes, I see that one process has finished sending and receiving all messages (10000, as specified in the command). The other process, has successfully sent all of its messages, but has only received a portion of them. In the run I have on my screen right now, out of 10,000 messages, the node is stuck only having received 5790 of them.

I believe this has to do with the handling of buffered messages in sockets when the FI_MULTI_RECV buffers in GASNet run out of space. The first indication of this is the log messages I get. Sometimes (but not all the time), I get this log dump:
```
libfabric:sockets:cq:_sock_cq_write():181<warn> Not enough space in CQ                        
...            
libfabric:sockets:cq:_sock_cq_write():181<warn> Not enough space in CQ                        
libfabric:sockets:ep_data:sock_rx_new_buffered_entry():109<warn> Exceeded buffered recv limit 
...
libfabric:sockets:ep_data:sock_rx_new_buffered_entry():109<warn> Exceeded buffered recv limit 
```

By default, GASNet allocates 8 multi-recv buffers, each 1MB in size. These are split between two receiving endpoints. If i greatly increase the size or number of the receive buffers (make each 8MB in size or post 8 times as many 1MB buffers) I can't reproduce this error. If I reduce the number of buffers to two (which is the minimum, as there are two MSG endpoints in GASNet/ofi) then the test won't function at all (and the test does function under these circumstances with other providers, like psm2).

### PR DESCRIPTION
**Pull request by epaulson10**
This bug was found by testing GASNet, which has some tests that very
rigourously stress messaging endpoints. The behaviour was that tests would
sporadically hang when put under heavy loads.

The root cause is found in the provider's handling of previously buffered
messages (ones that came in without a matching receive buffer). When the
handler grabbed a posted buffer to place a buffered message into, it was not
marking that receive buffer as free once it was done with it, which causes
buffers posted with FI_MULTI_RECV to become unusable in the future, which
leads to the provider running out of receive buffers.

After fixing the root cause, there was also an issue in calculating the
position in which to place a message in such a multi-recv buffer, which this
patch also fixes.

Signed-off-by: Erik Paulson <erik.r.paulson@intel.com>

### CODE REVIEW

**Comment by epaulson10** [comment]
Travis failed fi_multi_ep on the linux GCC build, but passed it on the Clang Linux build. Here was the output:
```
- name:   fi_multi_ep -e msg -p "sockets"

  result: Fail

  time:   90

  server_cmd: fi_multi_ep -e msg -p "sockets" -s 127.0.0.1

  server_stdout: |

    transmit(): common/shared.c:1950, ret=-11 (Resource temporarily unavailable)

  client_cmd: fi_multi_ep -e msg -p "sockets" -s 127.0.0.1 127.0.0.1

  client_stdout: |

    PASSED multi ep
```

This test passes when I run it on my Linux machine using GCC.

**Comment by shefty** [comment]
Ignore multi_ep tests. It fails frequently for any PR.  I haven't been able to isolate the cause of the failure.  The test itself looks like it should be okay, but it's a new test.

File: prov/sockets/src/sock_progress.c [hunk 1]
```diff
@@ -1166,8 +1166,11 @@ static int sock_pe_progress_buffered_rx(struct sock_rx_ctx *rx_ctx)
 		pe_entry.data_len = 0;
 		pe_entry.buf = 0L;
 		for (i = 0; i < rx_posted->rx_op.dest_iov_len && rem > 0; i++) {
-			if (used_len >= rx_posted->rx_op.dest_iov_len) {
-				used_len -= rx_posted->rx_op.dest_iov_len;
+			/* Try to find an appropriately sized iovec in the
+			 * posted array of iovecs. In the common case, there
+			 * is only one iovec, i.e. a single buffer */
```
Comment (shefty) [inline prov/sockets/src/sock_progress.c]: This comment is misleading.  The looping over the iovec is to find the first iovec entry where the data has not already been consumed.

Comment (epaulson10) [inline prov/sockets/src/sock_progress.c]: Okay, I updated it, copying almost exactly what you just said :)

**Comment by shefty** [comment]
thanks - also adding to v1.4.x branch

**Comment by epaulson10** [comment]
@shefty I think this needs to get picked to v1.5.x.

**Pull request merged**